File: src/test/java/no/nav/bidrag/beregn/forskudd/rest/consumer/SjablonConsumerTest.java
package no.nav.bidrag.beregn.forskudd.rest.consumer;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.when;

import java.util.List;
import no.nav.bidrag.beregn.forskudd.rest.TestUtil;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

@ExtendWith(MockitoExtension.class)
@SuppressWarnings("unchecked")
@DisplayName("SjablonConsumerTest")
class SjablonConsumerTest {

  @InjectMocks
  private SjablonConsumer sjablonConsumer;

  @Mock
  private RestTemplate restTemplateMock;

  @Test
  @DisplayName("Skal hente liste av sjablonverdier")
  void skalHenteListeAvSjablonverdier() {
    when(restTemplateMock.exchange(anyString(), eq(HttpMethod.GET), eq(null), (ParameterizedTypeReference<List<Sjablontall>>) any()))
        .thenReturn(new ResponseEntity<>(TestUtil.dummySjablonListe(), HttpStatus.OK));
    var sjablonResponse = sjablonConsumer.hentSjablontall();

    assertAll(
        () -> assertThat(sjablonResponse).isNotNull(),
        () -> assertThat(sjablonResponse.getBody().size()).isEqualTo(TestUtil.dummySjablonListe().size()),
        () -> assertThat(sjablonResponse.getBody().get(0).getTypeSjablon()).isEqualTo("0005")
    );
  }

  @Test
  @DisplayName("Skal returnere null")
  void skalReturnereNull() {
    when(restTemplateMock.exchange(anyString(), eq(HttpMethod.GET), eq(null), (ParameterizedTypeReference<List<Sjablontall>>) any()))
        .thenReturn(null);
    var sjablonResponse = sjablonConsumer.hentSjablontall();

    assertThat(sjablonResponse).isNull();
  }
}

File: src/test/kotlin/no/nav/bidrag/beregn/forskudd/rest/consumer/SjablonConsumerTest.kt
package no.nav.bidrag.beregn.forskudd.rest.consumer

import no.nav.bidrag.beregn.forskudd.rest.TestUtil
import no.nav.bidrag.beregn.forskudd.rest.exception.SjablonConsumerException
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.Assertions.assertAll
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyString
import org.mockito.ArgumentMatchers.eq
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.RestTemplate

@ExtendWith(MockitoExtension::class)
@SuppressWarnings("unchecked")
@DisplayName("SjablonConsumerTest")
internal class SjablonConsumerTest {
    @InjectMocks
    private lateinit var sjablonConsumer: SjablonConsumer

    @Mock
    private lateinit var restTemplateMock: RestTemplate

    @Test
    @DisplayName("Skal hente liste av Sjablontall når respons fra tjenesten er OK")
    fun skalHenteListeAvSjablontallNaarResponsFraTjenestenErOk() {
        `when`(
            restTemplateMock.exchange(
                anyString(),
                eq(HttpMethod.GET),
                eq(null),
                any<ParameterizedTypeReference<List<Sjablontall>>>(),
            ),
        )
            .thenReturn(ResponseEntity(TestUtil.dummySjablonSjablontallListe(), HttpStatus.OK))

        val sjablonResponse = sjablonConsumer.hentSjablonSjablontall()

        assertAll(
            { assertThat(sjablonResponse).isNotNull() },
            { assertThat(sjablonResponse.responseEntity.statusCode).isNotNull() },
            { assertThat(sjablonResponse.responseEntity.statusCode).isEqualTo(HttpStatus.OK) },
            { assertThat(sjablonResponse.responseEntity.body).isNotNull() },
            { assertThat(sjablonResponse.responseEntity.body?.size).isEqualTo(TestUtil.dummySjablonSjablontallListe().size) },
            {
                assertThat(
                    sjablonResponse.responseEntity.body?.get(0)?.typeSjablon,
                ).isEqualTo(TestUtil.dummySjablonSjablontallListe()[0].typeSjablon)
            },
        )
    }

    @Test
    @DisplayName("Skal kaste SjablonConsumerException når respons fra tjenesten ikke er OK for Sjablontall")
    fun skalKasteRestClientExceptionNaarResponsFraTjenestenIkkeErOkForSjablontall() {
        `when`(restTemplateMock.exchange(anyString(), eq(HttpMethod.GET), eq(null), any<ParameterizedTypeReference<List<Sjablontall>>>()))
            .thenThrow(HttpClientErrorException(HttpStatus.INTERNAL_SERVER_ERROR))
        assertThatExceptionOfType(SjablonConsumerException::class.java).isThrownBy { sjablonConsumer.hentSjablonSjablontall() }
    }
}
